results/pkg/api/server/server.go

```diff
@@ -157,51 +168,71 @@ func (s *Server) ListResultsResult(ctx context.Context, req *pb.ListResultsReque
 	}
-	// get all results from database
-	rows, err := s.db.Query("SELECT data FROM results")
-	if err != nil {
-		log.Printf("failed to query on database: %v", err)
-		return nil, status.Errorf(codes.Internal, "failed to query results: %v", err)
+
+	// checks and refines the pageSize
+	pageSize := req.GetPageSize()
+	if pageSize < 0 {
+		return nil, status.Error(codes.InvalidArgument, "PageSize should be greater than 0")
+	} else if pageSize == 0 {
+		pageSize = listResultsDefaultPageSize
+	} else if pageSize > listResultsMaximumPageSize {
+		pageSize = listResultsMaximumPageSize
 	}
-	var results []*pb.Result
-	for rows.Next() {
-		var b []byte
-		if err := rows.Scan(&b); err != nil {
-			log.Printf("failed to scan a row in query results: %v", err)
-			return nil, status.Errorf(codes.Internal, "failed to read result data: %v", err)
+
+	// retrieve the ListPageIdentifier from PageToken
+	var pageIdentifier *pb.ListPageIdentifier
+	pageToken := req.GetPageToken()
+	if pageToken != "" {
+		var err error
+		if pageIdentifier, err = decodePageToken(pageToken); err != nil {
+			return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("invalid PageToken: %v", err))
 		}
-		r := &pb.Result{}
-		if err := proto.Unmarshal(b, r); err != nil {
-			log.Printf("unmarshaling error: %v", err)
-			return nil, status.Errorf(codes.Internal, "failed to parse result data: %v", err)
+		if req.GetFilter() != pageIdentifier.GetFilter() {
+			return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("use a different CEL `filter` from the last page."))
 		}
-		results = append(results, r)
 	}
 
+	var prg cel.Program
+	var results []*pb.Result
+	var err error
 	// return all results back to users if empty query is given
-	if req.GetFilter() == "" {
-		return &pb.ListResultsResponse{Results: results}, nil
+	if req.GetFilter() != "" {
+		// filter from all results
+		prg, err = s.env.Program(ast)
+		if err != nil {
+			log.Printf("program construction error: %s", err)
+			return nil, status.Errorf(codes.InvalidArgument, "Error occurred during filter checking step, no Results found for the query string due to invalid field, invalid function to evaluate filter or missing double quotes around field value, please try to enter a query with correct type again: %v", err)
+		}
 	}
 
-	// filter from all results
-	prg, err := s.env.Program(ast)
-	if err != nil {
-		log.Printf("program construction error: %s", err)
-		return nil, status.Errorf(codes.InvalidArgument, "Error occurred during filter checking step, no Results found for the query string due to invalid field, invalid function to evaluate filter or missing double quotes around field value, please try to enter a query with correct type again: %v", err)
-	}
-	var resp []*pb.Result
-	for _, r := range results {
-		if ok, err := matchCelFilter(r, prg); err != nil {
-			return nil, err
-		} else if ok {
-			resp = append(resp, r)
+	var tx *sql.Tx
+	if tx, err = s.db.Begin(); err != nil {
+		return nil, err
+	}
```

Comment:
This is a case where it's more idiomatic to not use the inline if syntax - rule of thumb is to use it when the used values are only needed within the if.

```suggestion
	tx, err := s.db.Begin()
	if err != nil {
		return nil, err
	}
```